integration/volume_copy_up_test.go

```diff
@@ -144,34 +215,46 @@ func TestVolumeOwnership(t *testing.T) {
 	assert.Equal(t, expectedContainerOutput, string(stdout))
 
 	t.Logf("Check ownership of test directory on the host")
-	volumePaths, err := getHostPathForVolumes(*criRoot, cn)
+	// volumePaths, err := getHostPathForVolumes(*criRoot, cn)
+	volumePaths, err := getContainerVolumes(t, *criRoot, cn)
 	require.NoError(t, err)
 	assert.Equal(t, len(volumePaths), 1, "expected exactly 1 volume")
 
-	output, err := getOwnership(volumePaths[0])
+	output, err := getOwnership(volumePaths[volumePath])
 	require.NoError(t, err)
 	assert.Equal(t, expectedHostOutput, output)
 }
 
-func getHostPathForVolumes(criRoot, containerID string) ([]string, error) {
-	hostPath := filepath.Join(criRoot, "containers", containerID, "volumes")
-	if _, err := os.Stat(hostPath); err != nil {
-		return nil, err
-	}
-
-	volumes, err := os.ReadDir(hostPath)
-	if err != nil {
-		return nil, err
+func getContainerVolumes(t *testing.T, criRoot, containerID string) (map[string]string, error) {
+	client, err := RawRuntimeClient()
+	require.NoError(t, err, "failed to get raw grpc runtime service client")
+	request := &v1.ContainerStatusRequest{
+		ContainerId: containerID,
+		Verbose:     true,
 	}
+	response, err := client.ContainerStatus(context.TODO(), request)
+	require.NoError(t, err)
+	ret := make(map[string]string)
 
-	if len(volumes) == 0 {
-		return []string{}, nil
-	}
+	mounts := struct {
+		RuntimeSpec struct {
+			Mounts []specs.Mount `json:"mounts"`
+		} `json:"runtimeSpec"`
+	}{}
 
-	volumePaths := make([]string, len(volumes))
-	for idx, volume := range volumes {
-		volumePaths[idx] = filepath.Join(hostPath, volume.Name())
+	info := response.Info["info"]
+	err = json.Unmarshal([]byte(info), &mounts)
+	require.NoError(t, err)
+	containerVolumesHostPath := filepath.Join(criRoot, "containers", containerID, "volumes")
+	for _, mount := range mounts.RuntimeSpec.Mounts {
+		// norm, err := getFinalPath(mount.Source)
+		// require.NoError(t, err)
+		if runtime.GOOS == "linux" && strings.HasPrefix(mount.Source, containerVolumesHostPath) {
+			ret[mount.Destination] = mount.Source
+		} else if runtime.GOOS == "windows" {
+			// On windows we only get the volume mounts, so no need to do prefix checks.
+			ret[mount.Destination] = mount.Source
+		}
```

Comment:
This falls through completely if not Linux or Windows. It'd probably be clearer if one is picked as the "default" behaviour in case, e.g., the MacOS code ever lands.

Does it really matter if we don't do the prefix check on Linux? Does `mounts.RuntimeSpec.Mounts` end up with multiple entries for our desired mount-points (which would suggest a different issue to me...), or was this just so the `assert.Equalf(t, len(expectedVolumes), len(volumeMappings)` test above works?

If the latter, perhaps that test could instead check only that we have volume mounts for each expected volume, so that it's more robust against magic mounts on any platform. I note that the other caller of this function would need similar changes, so maybe that's a rainy-afternoon-with-nothing-else-particular-to-do idea at best.

The main advantage of that change is that we don't need to hard-code the `containerVolumesHostPath` algorithm in the test suite, in case someone refactors that logic to be more like Docker, which creates anonymous volumes to hold this data instead. (And the latter's probably a better approach... Containers should probably not be able to fill the disk where containerd's internal state is stored, bypassing anything the admin has done in the snapshotter configuration to isolate contained process disk usage. This is in parens because it's _definitely_ a future-discussion prompt.)

Reply:
> I can check that all expected volumes are there if we're okay with not checking for any unexpected volumes.

My _current opinion_ is that checking for unexpected volumes is more fragile than its value, given that we need to rely on non-API knowledge to identify them.

Reply:
Fair enough. Changing.